### PR TITLE
[AIRFLOW-5359] Update type annotations in BaseOperator

### PR DESCRIPTION
**Pull request by TobKed**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5359
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

**Comment by TobKed** [comment]
CC: @ashb

**Review by turbaszek — APPROVED** [review]

File: airflow/models/baseoperator.py
Comment (kaxil) [inline airflow/models/baseoperator.py]: This should be `Optional[Callable]`, all the callbacks are optional so they can take None or a callable function

File: airflow/models/baseoperator.py
Comment (kaxil) [inline airflow/models/baseoperator.py]: Same here

File: airflow/models/baseoperator.py
Comment (TobKed) [reply airflow/models/baseoperator.py]: I've raised question about `Optional` type on Airflow Slack [here](https://apache-airflow.slack.com/archives/CCPRP7943/p1567155425013200).
It was pointed there that arguments with a `None` default value are having an implicit `Optional[...]` type. It is default `mypy` behavior until it is not configured differently. Documentation about it here: [The mypy command line - None and Optional handling](https://mypy.readthedocs.io/en/stable/command_line.html#none-and-optional-handling) and here: [The mypy configuration file - None and Optional handling](https://mypy.readthedocs.io/en/stable/config_file.html#none-and-optional-handling).
Due to this I assumed less code is better and I wanted to make annotations to be more consistent (some of them have `Optional` and some not). I had the same approach in another PR [[AIRFLOW-5360] Type annotations for BaseSensorOperator](https://github.com/apache/airflow/pull/5966/) and some future ones on which I am currently working.

Comment (kaxil) [reply airflow/models/baseoperator.py]: Hmm. I am in favor of having Optional even if we have None as default.
But again that is my preference. 

I will leave it to other committers @mik-laj @potiuk

Comment (potiuk) [reply airflow/models/baseoperator.py]: There is a huge discussion here: https://github.com/python/typing/issues/275

Comment (kaxil) [reply airflow/models/baseoperator.py]: Thanks @potiuk for that link.

>Following discussion in python/typing#275, there is a consensus that it is better to require optional types to be made explicit. This PR changes the wording of PEP 484 to allow, but not require, type checkers to treat a None default as implicitly making an argument Optional.

So please let's not change this

Comment (kaxil) [reply airflow/models/baseoperator.py]: https://github.com/python/peps/pull/689 for you reference @TobKed

Comment (potiuk) [reply airflow/models/baseoperator.py]: Yeah. I read a bit more in detail and actually what I think now (I changed my mind) since there is this "trend" and possible future breaking change to disallow implicit Optional, maybe we should do more than that - namely we should convert all implicit Optionals and add '--disable-implicit-optional' flag to mypy and prevent anyone from ever adding implicit Optional?

Comment (TobKed) [reply airflow/models/baseoperator.py]: Thanks @kaxil and @potiuk. I've read all your references and I've changed my mind as well. I will change it accordingly. Changing `no_implicit_optional` mypy flag seems good way to go.

**Comment by TobKed** [comment]
@kaxil @potiuk I've applied changes.

**Comment by TobKed** [comment]
cc @mik-laj @potiuk

**Review by potiuk — APPROVED** [review]

**Pull request merged**
